src/core/utils/debug_uart.h

```diff
@@ -0,0 +1,186 @@
+#define PLATFORM_DEBUG_UART_H
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a very basic debug uart useful for logging
+ *   messages when the primary uart is already engaged (example the NCP)
+ *   and encapsulating logs is not practical.
+ *
+ * Implimentation Notes:
+ *
+ * Only 3 functions are required to be implimented by the platform.
+ * see/search for the word MUST in the comments below.
+ *
+ * The other functions (without MUST in the comment) are WEAK symbols
+ * that can optionally be overridden by the platform or the application.
+ *
+ * many of the other functions are simply convience functions to
+ * aid a developer during their normal course of work and are not
+ * intended to be present, or used in production system.
+ */
+
+#if __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Standard printf() to the debug uart with no log decoration.
+     *
+     * @param fmt[in] printf formatter text
+     *
+     * This is a debug convience function that is not intended to be
+     * used in anything other then "debug scenarios" by a developer.
+     *
+     * lf -> cr/lf mapping is automatically handled via otPlatDebugUart_putchar()
+     *
+     * @sa otPlatDebugUart_vprintf() for limitations
+     *
+     * This is a WEAK symbol that can easily be overridden as needed.
+     */
+    void otPlatDebugUart_printf(const char *fmt, ...);
```

Comment:
Are these function intended to be used in main/core openthread source file to log info?

If so, why not use the existing `otLog<Level><Module>()` APIs and just map `otPlatLog` to use a sperate debug/log UART (which may be separate channel/uart from what NCP uses)... 

Can you please elaborate on advantage/use-case of these function? Thanks.